### PR TITLE
[ticket/16389] Overhaul coding guidelines

### PR DESCRIPTION
**Pull request by gijsmartens**
PHPBB-16389

Checklist:

- [ ] Correct branch: master for new features; 3.3.x & 3.2.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html), [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [ ] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-16389

---

The coding guidelines are due for an overhaul.
There are currently (atleast in my eyes) a few issues:
- The guidelines are difficult to read/follow.
- The guidelines sections are difficult to link to.
- The guidelines are severely outdated: 10-15 years old!
- There is no code highlighting at all.
Which _(in my opinion)_ is weird when talking about specific code examples,
and makes the points less obvious and more difficult to see and understand.

---

#### Actual file changes
- Only _really_ changed the coding guidelines (not **Translations**, **Encodings**, etc..)
- Fixed most of the indentation so it is properly outlined.
- Fixed some of the wrapping, so it can actually be maintained.
- Added automatic numbering for all sections and sub-sections.
- Added anchors to each section: `#chapter1.section2.sub_section3`
- Enhanced the section headers to better distinguish the sections.
- Created a two column layout _(which is responsive)_.

---

#### Guidelines changes
- Barely made any chances to the existing guidelines.
- Made the rules more concise, rather than paragraphs of text.
- Added a few guidelines which were already followed but not documented.
  - Think of service naming conventions and their file placements.
  - Or exception naming and file placements.
- Replaced a part of the `SQL` section, 
  - Removed a part which was more a documentation on how the functions worked.
Which there already is proper documentation for, and a new PR that enhances it.
  - Added actual, concise rules. Which were already present but hidden in paragraphs of text.
- Added code highlighting, so code fragments actually make more sense rather than green blurs.

[Preview the complete overhauled guidelines](https://nimbusweb.me/s/image/3926906/rrb6esc7sl6fwb5byooq/a7SSSwlMbbu9OM31/anQn3iwOm0L3NVOA/screenshot-localhost-2020.03.03-22_08_13.png)
Or use this link: https://nimb.ws/uIoRFc

---

#### Suggestions
I am open to suggestions, by all means, let me know what you think.
That being said, I do think we need to start somewhere, as currently the guidelines as they are, are rather unmaintainable and outdated. Even if some rules are up for debate, I think we have to start off somewhere. Atleast with this we have a solid avenue and up-to-date document to work with.

---

#### CDN
Only thing I am not completely sure about, is that currently the code highlighting script and style ([HighlightJS](https://highlightjs.org/), which is also used by the s9e TextFormatter) is included through a CDN. I do not know if that is okay, or if a local copy should be used? And if so, how that should be done correclty?

### CODE REVIEW

**Comment by mrgoldy** [comment]
The following sections should probably also be looked at after this.
I haven't edited those in this PR, only updated the headers and some indentation / wrapping.
- Styling
- Templating
- Character sets & Encoding
- Translations

**Comment by hanakin** [comment]
Not trying to dismiss all the work you have put in. But I am not sure we really need this at all anymore. It just adds yet another place we need to update this stuff. Its better IMHO to just provide links to the actual documentation via a TOC file. Its also not in any sense of the word up to date or matching the development docs.

**Comment by mrgoldy** [comment]
Well that's not completely true. There is a distinct difference between documentation and guidelines.
These guidelines exist nowhere else, they are referenced from the documentation.
It's not "how things work", but "how things should be used".

I am fine with converting these guidelines into rST and including them in the documentation, that's fine by me. And that it does not match the Read-the-Docs is not completely true anymore, especially now I've created concise lists, code highlighting and a two column layout. I would even argue it might be better.

I was wondering the same thing myself aswell, why they are not part of the docs. I believe I figured out the reason, but can't think of it right now.

File: phpBB/docs/coding-guidelines.html [hunk 1]
```diff
@@ -33,1154 +44,1797 @@ <h1>Coding Guidelines</h1>
...
-		<div class="content">
-
-<a name="editorsettings"></a><h3>1.i. Editor Settings</h3>
-
-	<h4>Tabs vs Spaces:</h4>
-	<p>In order to make this as simple as possible, we will be using tabs, not spaces. We enforce 4 (four) spaces for one tab - therefore you need to set your tab width within your editor to 4 spaces. Make sure that when you <strong>save</strong> the file, it's saving tabs and not spaces. This way, we can each have the code be displayed the way we like it, without breaking the layout of the actual files.</p>
-	<p>Tabs in front of lines are no problem, but having them within the text can be a problem if you do not set it to the amount of spaces every one of us uses. Here is a short example of how it should look like:</p>
-
-	<div class="codebox"><pre>
-{TAB}$mode{TAB}{TAB}= $request->variable('mode', '');
```
Comment (dark-1) [inline phpBB/docs/coding-guidelines.html]: Where is this line gone ??
this explains better...

**Comment by hanakin** [comment]
all this css will need to be run through stylelint

**Pull request closed without merging**